### PR TITLE
8296956: [JVMCI] HotSpotResolvedJavaFieldImpl.getIndex returns wrong value

### PR DESCRIPTION
**Pull request by dougxc**
This PR fixes a bug related to `HotSpotResolvedJavaFieldImpl.index`. Its value is passed into the `HotSpotResolvedJavaFieldImpl` constructor as an `int`, and is returned by `getIndex()` as an `int` but it was stored as a `short`. This meant that unsigned 16-bit values were not handled correctly.

Also included are some related JVMCI cleanups:
* added and fixed doc related to `ResolvedJavaField.getOffset()`
* replaced assertions with always-enabled checks

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8296956](https://bugs.openjdk.org/browse/JDK-8296956): [JVMCI] HotSpotResolvedJavaFieldImpl.getIndex returns wrong value

### Reviewers
 * [Tobias Hartmann](https://openjdk.org/census#thartmann) (@TobiHartmann - **Reviewer**)
 * [Tom Rodriguez](https://openjdk.org/census#never) (@tkrodriguez - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/11142/head:pull/11142` \
`$ git checkout pull/11142`

Update a local copy of the PR: \
`$ git checkout pull/11142` \
`$ git pull https://git.openjdk.org/jdk pull/11142/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 11142`

View PR using the GUI difftool: \
`$ git pr show -t 11142`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/11142.diff">https://git.openjdk.org/jdk/pull/11142.diff</a>

</details>

### CODE REVIEW

File: src/jdk.internal.vm.ci/share/classes/jdk.vm.ci.hotspot/src/jdk/vm/ci/hotspot/HotSpotConstantPool.java [hunk 1]
```diff
@@ -256,11 +256,23 @@ private static int rawIndexToConstantPoolCacheIndex(int rawIndex, int opcode) {
...
-                            opcode == Bytecodes.INVOKEVIRTUAL || opcode == Bytecodes.INVOKESPECIAL || opcode == Bytecodes.INVOKESTATIC : "unexpected invoke opcode " + opcode;
-            index = rawIndex + config().constantPoolCpCacheIndexTag;
+            if (opcode == Bytecodes.GETFIELD ||
+                            opcode == Bytecodes.PUTFIELD ||
+                            opcode == Bytecodes.GETSTATIC ||
+                            opcode == Bytecodes.PUTSTATIC ||
+                            opcode == Bytecodes.INVOKEINTERFACE ||
+                            opcode == Bytecodes.INVOKEVIRTUAL ||
+                            opcode == Bytecodes.INVOKESPECIAL ||
+                            opcode == Bytecodes.INVOKESTATIC) {
```
Comment (TobiHartmann) [inline src/jdk.internal.vm.ci/share/classes/jdk.vm.ci.hotspot/src/jdk/vm/ci/hotspot/HotSpotConstantPool.java:270]: Indentation looks a bit off.

**Review by TobiHartmann — APPROVED** [review]
Looks good to me otherwise.

File: src/jdk.internal.vm.ci/share/classes/jdk.vm.ci.hotspot/src/jdk/vm/ci/hotspot/HotSpotConstantPool.java (hunk 1 above)
Comment (dougxc) [reply src/jdk.internal.vm.ci/share/classes/jdk.vm.ci.hotspot/src/jdk/vm/ci/hotspot/HotSpotConstantPool.java:270]: That's due to the Eclipse formatter enforced style adopted across JVMCI and Graal. I can add comments to explicitly disable it here if you want but I'm not sure it's worth it.

Comment (TobiHartmann) [reply src/jdk.internal.vm.ci/share/classes/jdk.vm.ci.hotspot/src/jdk/vm/ci/hotspot/HotSpotConstantPool.java:270]: I see. No, that's not worth it then.

**Pull request closed without merging**